In [3]:
import pandas as pd

In [4]:
import numpy as np

In [5]:
import wget

## EDA

### Load the data. - See below. 
    
### Look at the price variable. Does it have a long tail? 
### Answer : Yes, it has a long tail.  scroll further for this.


In [6]:
data2 = 'https://raw.githubusercontent.com/alexeygrigorev/datasets/master/AB_NYC_2019.csv'

In [7]:
ny_airbnb = wget.download(data2)

100% [......................................................] 7077973 / 7077973

In [8]:
ny_abnb = pd.read_csv(ny_airbnb)

In [9]:
ny_abnb.head()

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

In [10]:
ny_abnb.columns.str.lower

<bound method StringMethods.lower of <pandas.core.strings.accessor.StringMethods object at 0x122684d30>>

In [11]:
ny_abnb.head(2)

id                                name  host_id host_name  \
0  2539  Clean & quiet apt home by the park     2787      John   
1  2595               Skylit Midtown Castle     2845  Jennifer   

  neighbourhood_group neighbourhood  latitude  longitude        room_type  \
0            Brooklyn    Kensington  40.64749  -73.97237     Private room   
1           Manhattan       Midtown  40.75362  -73.98377  Entire home/apt   

   price  minimum_nights  number_of_reviews last_review  reviews_per_month  \
0    149               1                  9  2018-10-19               0.21   
1    225               1                 45  2019-05-21               0.38   

   calculated_host_listings_count  availability_365  
0                               6               365  
1                               2               355

In [12]:
ny_abnb.columns = ny_abnb.columns.str.replace(' ','_')

In [13]:
ny_abnb.head(2)

id                                name  host_id host_name  \
0  2539  Clean & quiet apt home by the park     2787      John   
1  2595               Skylit Midtown Castle     2845  Jennifer   

  neighbourhood_group neighbourhood  latitude  longitude        room_type  \
0            Brooklyn    Kensington  40.64749  -73.97237     Private room   
1           Manhattan       Midtown  40.75362  -73.98377  Entire home/apt   

   price  minimum_nights  number_of_reviews last_review  reviews_per_month  \
0    149               1                  9  2018-10-19               0.21   
1    225               1                 45  2019-05-21               0.38   

   calculated_host_listings_count  availability_365  
0                               6               365  
1                               2               355

In [14]:
ny_abnb.dtypes

id                                  int64
name                               object
host_id                             int64
host_name                          object
neighbourhood_group                object
neighbourhood                      object
latitude                          float64
longitude                         float64
room_type                          object
price                               int64
minimum_nights                      int64
number_of_reviews                   int64
last_review                        object
reviews_per_month                 float64
calculated_host_listings_count      int64
availability_365                    int64
dtype: object

In [15]:
cols = ["latitude", 'longitude','price','minimum_nights','number_of_reviews','reviews_per_month','calculated_host_listings_count','availability_365']

In [16]:
ny_abnb_sub = ny_abnb[cols]

In [17]:
ny_abnb_sub.head()

latitude  longitude  price  minimum_nights  number_of_reviews  \
0  40.64749  -73.97237    149               1                  9   
1  40.75362  -73.98377    225               1                 45   
2  40.80902  -73.94190    150               3                  0   
3  40.68514  -73.95976     89               1                270   
4  40.79851  -73.94399     80              10                  9   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

In [18]:
ny_abnb_sub.shape

(48895, 8)

In [19]:
ny_abnb_sub.dtypes

latitude                          float64
longitude                         float64
price                               int64
minimum_nights                      int64
number_of_reviews                   int64
reviews_per_month                 float64
calculated_host_listings_count      int64
availability_365                    int64
dtype: object

In [20]:
ny_abnb_sub.isnull().sum()

latitude                              0
longitude                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64

## Question 1: 
### Find a feature with missing values. How many missing values does it have?
### Answer : The  feature with missing value is "reviews_per_month"with 10052 missing values of total 48895 rows. 

In [21]:
for col in ny_abnb_sub.columns:
    print(col)
    print(ny_abnb_sub[col].unique()[:5])
    print(ny_abnb_sub[col].nunique())
    print()

latitude
[40.64749 40.75362 40.80902 40.68514 40.79851]
19048

longitude
[-73.97237 -73.98377 -73.9419  -73.95976 -73.94399]
14718

price
[149 225 150  89  80]
674

minimum_nights
[ 1  3 10 45  2]
109

number_of_reviews
[  9  45   0 270  74]
394

reviews_per_month
[0.21 0.38  nan 4.64 0.1 ]
937

calculated_host_listings_count
[6 2 1 4 3]
47

availability_365
[365 355 194   0 129]
366



In [22]:
ny_abnb_sub.describe().round(2)

latitude  longitude     price  minimum_nights  number_of_reviews  \
count  48895.00   48895.00  48895.00        48895.00           48895.00   
mean      40.73     -73.95    152.72            7.03              23.27   
std        0.05       0.05    240.15           20.51              44.55   
min       40.50     -74.24      0.00            1.00               0.00   
25%       40.69     -73.98     69.00            1.00               1.00   
50%       40.72     -73.96    106.00            3.00               5.00   
75%       40.76     -73.94    175.00            5.00              24.00   
max       40.91     -73.71  10000.00         1250.00             629.00   

       reviews_per_month  calculated_host_listings_count  availability_365  
count           38843.00                        48895.00          48895.00  
mean                1.37                            7.14            112.78  
std                 1.68                           32.95            131.62  
min                 0.01                            1.00              0.00  
25%                 0.19                            1.00              0.00  
50%                 0.72                            1.00             45.00  
75%                 2.02                            2.00            227.00  
max                58.50                          327.00            365.00

## Question 2: 
### What's the median (50% percentile) for variable 'minimum_nights'?
### Answer : The median (50% percentile) for variable 'minimum_nights' is 3.00 or 3 nights is minimum stay in airbnb. 

## EDA
### we also see here that price has a long tail since mean for price is 152.72 but max value is 10000.00, which shows that it has a long tail. 

In [23]:
import seaborn as sns

In [24]:
import matplotlib.pyplot as plt

In [25]:
%matplotlib inline

In [29]:
n = len(ny_abnb_sub)

n_val = int(n * 0.2)
n_test = int(n * 0.2)
n_train = n - n_val - n_test

In [30]:
n

48895

In [31]:
n_val, n_test, n_train

(9779, 9779, 29337)

In [32]:
ny_abnb_sub.iloc[[10, 0, 3, 5]]

latitude  longitude  price  minimum_nights  number_of_reviews  \
10  40.80316  -73.96545    135               5                 53   
0   40.64749  -73.97237    149               1                  9   
3   40.68514  -73.95976     89               1                270   
5   40.74767  -73.97500    200               3                 74   

    reviews_per_month  calculated_host_listings_count  availability_365  
10               0.43                               1                 6  
0                0.21                               6               365  
3                4.64                               1               194  
5                0.59                               1               129

In [43]:
idx = np.arange(n)

In [44]:
np.random.seed(42)
np.random.shuffle(idx)

In [45]:
df_train = ny_abnb_sub.iloc[idx[:n_train]]
df_val = ny_abnb_sub.iloc[idx[n_train:n_train+n_val]]
df_test = ny_abnb_sub.iloc[idx[n_train+n_val:]]

In [46]:
df_train.head(4)

latitude  longitude  price  minimum_nights  number_of_reviews  \
879    40.64354  -73.97777     89               3                 62   
44383  40.70666  -73.90779     30              21                  0   
15394  40.76116  -73.99016    120               2                 17   
43230  40.70763  -74.01050    470               2                  5   

       reviews_per_month  calculated_host_listings_count  availability_365  
879                 0.71                               1               189  
44383                NaN                               1                73  
15394               0.43                               1                 0  
43230               1.88                             327               272

In [47]:
len(df_train), len(df_val), len(df_test)

(29337, 9779, 9779)

In [48]:
df_train

latitude  longitude  price  minimum_nights  number_of_reviews  \
879    40.64354  -73.97777     89               3                 62   
44383  40.70666  -73.90779     30              21                  0   
15394  40.76116  -73.99016    120               2                 17   
43230  40.70763  -74.01050    470               2                  5   
16332  40.79658  -73.93287    199               2                 30   
...         ...        ...    ...             ...                ...   
14975  40.71178  -73.95779     95               2                  0   
14661  40.78638  -73.97004    101              30                 10   
24515  40.72716  -73.98524    100               3                 75   
31460  40.79578  -73.94523     68               2                 31   
29114  40.69671  -73.94896     45               1                  1   

       reviews_per_month  calculated_host_listings_count  availability_365  
879                 0.71                               1               189  
44383                NaN                               1                73  
15394               0.43                               1                 0  
43230               1.88                             327               272  
16332               0.80                               1                30  
...                  ...                             ...               ...  
14975                NaN                               1                 0  
14661               0.40                               1                 0  
24515               3.12                               1                60  
31460               2.10                               1               265  
29114               0.05                               1                 0  

[29337 rows x 8 columns]

In [49]:
df_train = df_train.reset_index(drop=True)
df_val = df_val.reset_index(drop=True)
df_test = df_test.reset_index(drop=True)

In [50]:
y_train = np.log1p(df_train.price.values)
y_val = np.log1p(df_val.price.values)
y_test = np.log1p(df_test.price.values)

In [51]:
y_train

array([4.49980967, 3.4339872 , 4.79579055, ..., 4.61512052, 4.2341065 ,
       3.8286414 ])

In [52]:
del df_train['price']
del df_val['price']
del df_test['price']

In [53]:
df_train.shape

(29337, 7)

In [54]:
y_train.shape

(29337,)

In [56]:
df_train.head(2)

latitude  longitude  minimum_nights  number_of_reviews  reviews_per_month  \
0  40.64354  -73.97777               3                 62               0.71   
1  40.70666  -73.90779              21                  0                NaN   

   calculated_host_listings_count  availability_365  
0                               1               189  
1                               1                73

In [57]:
df_train.describe().round(2)

latitude  longitude  minimum_nights  number_of_reviews  \
count  29337.00   29337.00        29337.00           29337.00   
mean      40.73     -73.95            6.99              23.31   
std        0.05       0.05           21.01              44.63   
min       40.51     -74.24            1.00               0.00   
25%       40.69     -73.98            1.00               1.00   
50%       40.72     -73.96            2.00               5.00   
75%       40.76     -73.94            5.00              24.00   
max       40.91     -73.71         1000.00             629.00   

       reviews_per_month  calculated_host_listings_count  availability_365  
count           23339.00                        29337.00          29337.00  
mean                1.37                            7.13            112.83  
std                 1.65                           32.84            131.73  
min                 0.01                            1.00              0.00  
25%                 0.19                            1.00              0.00  
50%                 0.71                            1.00             45.00  
75%                 2.03                            2.00            228.00  
max                27.95                          327.00            365.00

In [58]:
df_train.isnull().sum()

latitude                             0
longitude                            0
minimum_nights                       0
number_of_reviews                    0
reviews_per_month                 5998
calculated_host_listings_count       0
availability_365                     0
dtype: int64

In [69]:
df_train.columns

Index(['latitude', 'longitude', 'minimum_nights', 'number_of_reviews',
       'reviews_per_month', 'calculated_host_listings_count',
       'availability_365'],
      dtype='object')

In [70]:
base = ['latitude', 'longitude', 'minimum_nights', 'number_of_reviews',
       'reviews_per_month', 'calculated_host_listings_count',
       'availability_365']

In [79]:
X_train_zero = df_train.fillna(0).values

In [78]:
X_train_zero

array([[ 4.064354e+01, -7.397777e+01,  3.000000e+00, ...,  7.100000e-01,
         1.000000e+00,  1.890000e+02],
       [ 4.070666e+01, -7.390779e+01,  2.100000e+01, ...,  0.000000e+00,
         1.000000e+00,  7.300000e+01],
       [ 4.076116e+01, -7.399016e+01,  2.000000e+00, ...,  4.300000e-01,
         1.000000e+00,  0.000000e+00],
       ...,
       [ 4.072716e+01, -7.398524e+01,  3.000000e+00, ...,  3.120000e+00,
         1.000000e+00,  6.000000e+01],
       [ 4.079578e+01, -7.394523e+01,  2.000000e+00, ...,  2.100000e+00,
         1.000000e+00,  2.650000e+02],
       [ 4.069671e+01, -7.394896e+01,  1.000000e+00, ...,  5.000000e-02,
         1.000000e+00,  0.000000e+00]])

In [88]:
X_train_zero.shape

(29337, 7)

In [93]:
def train_linear_regression(X, y):
    ones = np.ones(X.shape[0])
    X = np.column_stack([ones, X])

    XTX = X.T.dot(X)
    XTX_inv = np.linalg.inv(XTX)
    w_full = XTX_inv.dot(X.T).dot(y)
    
    return w_full[0], w_full[1:]

In [81]:
w0, w = train_linear_regression(X_train_zero, y_train)

y_pred = w0 + X_train_zero.dot(w)

In [82]:
w0

-419.9126587296756

In [83]:
w

array([ 1.30676225e+00, -5.02167855e+00, -5.59803110e-04, -4.07634896e-04,
       -8.25470065e-03,  1.33371320e-03,  6.50203890e-04])

In [84]:
def rmse(y, y_pred):
    se = (y - y_pred) ** 2
    mse = se.mean()
    return np.sqrt(mse)

In [85]:
rmse(y_train, y_pred)

0.6428860484860434

In [86]:
X_train_mean = df_train.fillna(df_train.mean()).values

In [87]:
X_train_mean

array([[ 4.06435400e+01, -7.39777700e+01,  3.00000000e+00, ...,
         7.10000000e-01,  1.00000000e+00,  1.89000000e+02],
       [ 4.07066600e+01, -7.39077900e+01,  2.10000000e+01, ...,
         1.36854921e+00,  1.00000000e+00,  7.30000000e+01],
       [ 4.07611600e+01, -7.39901600e+01,  2.00000000e+00, ...,
         4.30000000e-01,  1.00000000e+00,  0.00000000e+00],
       ...,
       [ 4.07271600e+01, -7.39852400e+01,  3.00000000e+00, ...,
         3.12000000e+00,  1.00000000e+00,  6.00000000e+01],
       [ 4.07957800e+01, -7.39452300e+01,  2.00000000e+00, ...,
         2.10000000e+00,  1.00000000e+00,  2.65000000e+02],
       [ 4.06967100e+01, -7.39489600e+01,  1.00000000e+00, ...,
         5.00000000e-02,  1.00000000e+00,  0.00000000e+00]])

In [89]:
X_train_mean.shape

(29337, 7)

In [90]:
w0mean, w_mean = train_linear_regression(X_train_mean, y_train)

y_pred_mean = w0mean + X_train_zero.dot(w_mean)

In [91]:
rmse(y_train, y_pred_mean)

0.6431116056592563

## Question 3
### Which option gives better RMSE? 
### Solution : we see that both the RMSE values are around 0.643 which is equal. Therefore, filling with mean or zero gives similar results and both are equally good. 

In [95]:
def train_linear_regression_reg(X, y, r=0.001):
    ones = np.ones(X.shape[0])
    X = np.column_stack([ones, X])

    XTX = X.T.dot(X)
    XTX = XTX + r * np.eye(XTX.shape[0])

    XTX_inv = np.linalg.inv(XTX)
    w_full = XTX_inv.dot(X.T).dot(y)
    
    return w_full[0], w_full[1:]

In [97]:
df_val

latitude  longitude  minimum_nights  number_of_reviews  \
0     40.66431  -73.93216               3                 26   
1     40.70631  -73.92064               2                  2   
2     40.85244  -73.93126               2                 11   
3     40.72100  -73.99218               2                  5   
4     40.66943  -73.74778               2                 85   
...        ...        ...             ...                ...   
9774  40.58943  -73.81192               2                 27   
9775  40.80734  -73.94666               2                  4   
9776  40.83987  -73.94279               2                  1   
9777  40.78804  -73.95382               5                  4   
9778  40.71508  -73.97895               2                  7   

      reviews_per_month  calculated_host_listings_count  availability_365  
0                  0.33                               1               158  
1                  0.26                               1                 0  
2                  0.29                               2                16  
3                  0.74                               1               326  
4                  1.99                               5               333  
...                 ...                             ...               ...  
9774               0.72                               2                 0  
9775               0.10                               1                 0  
9776               0.03                               1                 0  
9777               0.11                               1                 0  
9778               2.88                               1                27  

[9779 rows x 7 columns]

In [98]:
for r in [0.0, 0.00001, 0.0001, 0.001, 0.1, 1, 10]:
    #X_train_ = prepare_X(X_train_zero)
    w0, w = train_linear_regression_reg(X_train_zero, y_train, r=r)

    #X_val = prepare_X(df_val)
    y_pred = w0 + df_val.dot(w)
    score = rmse(y_val, y_pred)
    
    print(r, w0, score)

0.0 -419.9126587296756 0.6148743913855588
1e-05 -419.41376377130945 0.6148722614878929
0.0001 -414.9764924318162 0.6148583737902735
0.001 -375.2736527028995 0.615138570503776
0.1 -32.56256055225585 0.646970607496538
1 -3.49921683772884 0.651951437046105
10 -0.3512767591669603 0.6524962776865758


## Question 4 
### Which r  [0, 0.000001, 0.0001, 0.001, 0.01, 0.1, 1, 5, 10] gives best RMSE ? 
### Solution: We see that r = 0.0001 and 0.001 both give almost same RMSE 

In [99]:
ny_abnb_sub

latitude  longitude  price  minimum_nights  number_of_reviews  \
0      40.64749  -73.97237    149               1                  9   
1      40.75362  -73.98377    225               1                 45   
2      40.80902  -73.94190    150               3                  0   
3      40.68514  -73.95976     89               1                270   
4      40.79851  -73.94399     80              10                  9   
...         ...        ...    ...             ...                ...   
48890  40.67853  -73.94995     70               2                  0   
48891  40.70184  -73.93317     40               4                  0   
48892  40.81475  -73.94867    115              10                  0   
48893  40.75751  -73.99112     55               1                  0   
48894  40.76404  -73.98933     90               7                  0   

       reviews_per_month  calculated_host_listings_count  availability_365  
0                   0.21                               6               365  
1                   0.38                               2               355  
2                    NaN                               1               365  
3                   4.64                               1               194  
4                   0.10                               1                 0  
...                  ...                             ...               ...  
48890                NaN                               2                 9  
48891                NaN                               2                36  
48892                NaN                               1                27  
48893                NaN                               6                 2  
48894                NaN                               1                23  

[48895 rows x 8 columns]

In [118]:
n

48895

In [125]:
np.random.seed(0)
np.random.shuffle(idx)

In [126]:
idx

array([43813, 32734, 25276, ..., 42613, 43567,  2732])

In [124]:
idx = np.arange(n)

In [103]:
df_train = ny_abnb_sub.iloc[idx[:n_train]]
df_val = ny_abnb_sub.iloc[idx[n_train:n_train+n_val]]
df_test = ny_abnb_sub.iloc[idx[n_train+n_val:]]

In [127]:
for s in [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]:
    #X_train_ = prepare_X(X_train_zero)
    idx = np.arange(n)
    np.random.seed(s)
    np.random.shuffle(idx)
    df_train = ny_abnb_sub.iloc[idx[:n_train]]
    df_val = ny_abnb_sub.iloc[idx[n_train:n_train+n_val]]
    df_test = ny_abnb_sub.iloc[idx[n_train+n_val:]]
    df_train = df_train.reset_index(drop=True)
    df_val = df_val.reset_index(drop=True)
    df_test = df_test.reset_index(drop=True)
    y_train = np.log1p(df_train.price.values)
    y_val = np.log1p(df_val.price.values)
    y_test = np.log1p(df_test.price.values)
    del df_train['price']
    del df_val['price']
    del df_test['price']
    X_train_zero = df_train.fillna(0).values
    
    w0, w = train_linear_regression_reg(X_train_zero, y_train, r=0.001)

    #X_val = prepare_X(df_val)
    y_pred = w0 + df_val.dot(w)
    score = rmse(y_val, y_pred)
    
    print(s, w0, score)

0 -381.58558620827785 0.622247594033539
1 -371.06410933504276 0.6143364210550257
2 -378.42452200757873 0.6217128931841098
3 -377.544642666665 0.6135950935332248
4 -376.93543819715427 0.6138238535962295
5 -383.4140285674481 0.6040152771731441
6 -382.10333688193697 0.6069119792470334
7 -382.040998459231 0.6232797269720479
8 -382.04860512093194 0.6216666727955438
9 -381.18844690872874 0.620019239697519


In [130]:
score_sd = np.array([0.6222, 0.614, 0.621, 0.613, 0.613,0.604, 0.606, 0.623,0.621, 0.620])

In [131]:
std_dev = np.std(score_sd)

In [132]:
std_dev

0.0064703632046431555

## Question 5 
### What's the standard deviation of all the scores?
### Std dev is 0.006 but the closest option to our solution is 0.008. 

In [133]:
idx = np.arange(n)

In [134]:
np.random.seed(9)
np.random.shuffle(idx)

In [135]:
df_train = ny_abnb_sub.iloc[idx[:n_train]]
df_val = ny_abnb_sub.iloc[idx[n_train:n_train+n_val]]
df_test = ny_abnb_sub.iloc[idx[n_train+n_val:]]

In [145]:
df_full_train = pd.concat([df_train, df_val])

In [146]:
df_full_train = df_full_train.reset_index(drop=True)

In [147]:
X_full_train = df_full_train.fillna(0).values

In [148]:
X_full_train

array([[ 4.070423e+01, -7.391595e+01,  1.000000e+00, ...,  1.300000e-01,
         1.000000e+00,  0.000000e+00],
       [ 4.077115e+01, -7.392275e+01,  4.000000e+01, ...,  2.700000e-01,
         1.000000e+00,  2.070000e+02],
       [ 4.070401e+01, -7.400932e+01,  2.900000e+01, ...,  0.000000e+00,
         9.600000e+01,  1.000000e+00],
       ...,
       [ 4.072982e+01, -7.398593e+01,  2.000000e+00, ...,  1.100000e-01,
         1.000000e+00,  0.000000e+00],
       [ 4.071969e+01, -7.395620e+01,  1.000000e+00, ...,  2.000000e-02,
         1.000000e+00,  0.000000e+00],
       [ 4.066949e+01, -7.392172e+01,  1.000000e+00, ...,  1.910000e+00,
         2.000000e+00,  3.230000e+02]])

In [149]:
X_full_train.shape

(39116, 7)

In [141]:
y_train = np.log1p(df_train.price.values)
y_val = np.log1p(df_val.price.values)
y_test = np.log1p(df_test.price.values)

In [142]:
y_full_train = np.concatenate([y_train, y_val])

In [144]:
del df_train['price']
del df_val['price']
del df_test['price']

In [150]:
w0, w = train_linear_regression_reg(X_train_zero, y_train, r=0.001)

In [151]:
X_test = df_test
y_pred = w0 + X_test.dot(w)
score = rmse(y_test, y_pred)
score

0.6164431461525183

## Question 6 

### What's the RMSE on test dataset? 

### RMSE is 0.61 so selecting the closest option which is 0.60 